### ISSUE TITLE
jpegPhoto LDAP Integration

### ISSUE THREAD

**Issue by asmartin** [gogits/gogs#2543]
Gogs looks like a fantastic git repository management system, and the LDAP support is great! To that end, would it be possible to add support for pulling avatars from the `jpegPhoto` attribute in LDAP? This would be ideal for my use case

**Comment by nanoant** [comment]
Looks very similar to #2488. The idea is nice, but the problem is about syncing this data with Gogs. Gogs would need to either have some extra bg process to sync avatars, or feed avatar directly from LDAP. But this technically could overload LDAP with "avatar" request on each page render.

I wonder if wouldn't be easier if Gogs would let you provide some URL template for an avatar image?

**Comment by asmartin** [comment]
Reading directly from LDAP would be best since then there's no need to store or update a cache of LDAP avatars locally. Once you do the LDAP query to get the other information about the user (e.g. first name and last name) if you also get the jpegPhoto attribute (and store it in the `$jpegPhoto` PHP variable), you can display it as an image as follows:
```
<img src="data:image/jpg;base64,$jpegPhoto"/>
```

I don't think you can access jpegPhoto over HTTP - as far as I am aware you have to perform an LDAP query to pull information out of LDAP, so for my use case having a URL template would not help.

**Comment by Unknwon** [comment]
> I don't think you can access jpegPhoto over HTTP - as far as I am aware you have to perform an LDAP query to pull information out of LDAP, so for my use case having a URL template would not help.

Yes I think Gogs still needs a bg process as @nanoant mentioned, to periodically fetching the information.

**Comment by Lafriks** [comment]
It's related to #1268 where bg process is also required

**Comment by DMWOO7** [comment]
I think some kind of background process for keep users in sync would be a real additional value. For this issue to avoid the overhead of manual sync those avatars. My goal is to use LDAP als central user management system, so that our users maintain their data only there, and all connected applications like gogs would fetch their own data from there. 

By some kind of sync job, we could also [fix the known problem of changed permission, which are currently not automatically fetched by gogs](https://github.com/gogits/gogs/issues/2855). A good example would be the atlassian tools like Jira or Confluence: They use a background job, which syncs all users once a hour. All information is updated, not only attribute, also group member ships. 

Additionally, they fetch information from LDAP on each login. If they're different from the local ones, the local ones get updated. In my view, that's a good compromise between performance (avoid the overhead of a LDAP request on every page load) and keeping local user data in sync. Cause you've a maximum delay of 1 hour in the worst case, when something is updated in LDAP. If we need the changes earlier, we can simple tell the user to logout and login again, to take effect of recent LDAP changes like additional groups for extended permission.

**Comment by motlib** [comment]
I have seen the same feature implemented in the kanboard project (https://github.com/kanboard/kanboard). In that project the avatar picture is loaded from LDAP, when a user logs in and then cached as a file on the server. If you want to update the avatar picture, you need to delete the cached picture from your profile settings. Then it will be loaded again from LDAP on next login. See https://kanboard.net/documentation/ldap-profile-picture for a short description. 

From my experience this is a simple solution that works reasonably well without a background process.

**Comment by sidescraper** [comment]
That would be great

**Comment by kbrowder** [comment]
Sorry to revive but couldn't this just sync on user login?  Eg update the users photo each time the user logs in?  Obviously it'd be great to have a background process update the photo for everyone but maybe this would be a close enough approximation?
